kornia/contrib/kmeans.py

```diff
@@ -0,0 +1,172 @@
+# based on https://github.com/subhadarship/kmeans_pytorch
+
+from __future__ import annotations
+
+import torch
+
+from kornia.core import Tensor
+from kornia.core.check import KORNIA_CHECK, KORNIA_CHECK_SHAPE
+
+
+class KMeans:
+    """Implements the kmeans clustering algorithm with euclidean distance as similarity measure.
+
+    Args:
+        num_clusters: number of clusters the data has to be assigned to
+        cluster_centers: tensor of starting cluster centres can be passed instead of num_clusters
+        tolerance: float value. the algorithm terminates if the shift in centers is less than tolerance
+        max_iterations: number of iterations to run the algorithm for
+        device: the device to which all tensors are moved to and computed
+        seed: number to set torch manual seed for reproducibility
+
+    Example:
+    >>> kmeans = kornia.contrib.KMeans(3, None, 10e-4, 100, device, 0)
+    >>> kmeans.fit(torch.rand((1000, 5))
+    >>> predictions = kmeans.predict(torch.rand((10, 5)))
+    """
+
+    def __init__(
+        self,
+        num_clusters: int,
+        cluster_centers: Tensor,
+        tolerance: float = 10e-4,
+        max_iterations: int = 0,
+        seed: int | None = None,
+    ) -> None:
+        KORNIA_CHECK(num_clusters != 0, "num_clusters can't be 0")
+
+        # cluster_centers should have only 2 dimensions
+        if cluster_centers is not None:
+            KORNIA_CHECK_SHAPE(cluster_centers, ["C", "D"])
+
+        self.num_clusters = num_clusters
+        self.cluster_centers = cluster_centers
+        self.tolerance = tolerance
+        self.max_iterations = max_iterations
+
+        self.final_cluster_assignments = None
+        self.final_cluster_centers = None
```

Comment:
If you need to initialize this variable, then you need to add a typing here, otherwise, the type checker will be mad because when it are updated to Tensor.

```suggestion
        self.final_cluster_assignments: None | Tensor = None
        self.final_cluster_centers: None | Tensor = None
```

Reply:
also, i advocate to maybe go with internal/private member for these